### PR TITLE
[feat] - supervised launchd LaunchAgent for gate.py/harness (opt-in, highest risk)

### PR DESCRIPTION
**Pull request by cgfixit**
## ⚠️ Read this before reviewing the diff

#911

This PR is explicitly the **highest-risk** item of the four-part "Next integrations" series (`docs/work/MACOS_LAUNCHD_INTEGRATION_PLAN.md`, from #908). It makes gate.py or the coding harness capable of running as an **always-on, auto-restarting background network listener** — surviving logout, reboot, and crashes — instead of only running while an operator has a terminal open. That is a genuine change in this machine's availability/security posture, not just a convenience. The planning doc that requested this integration said so itself: *"deserves its own threat-model review before implementation, not a bundled add-on to a scheduler PR."*

Nothing in this PR is auto-loaded or wired to run automatically — see below for exactly what ships and what stays a deliberate, separate operator action.

## Branch naming
`claude/macos-gate-harness-launchagent` — Claude Code driver, matches the required `claude/<feature>` prefix.

## Title
`[feat] - supervised launchd LaunchAgent for gate.py/harness (opt-in, highest risk)`

---

## Proposed changes

Adds `macos/generate_service_plist.py` — a standalone script (Darwin-only, no CyClaw package import beyond the stdlib-only `utils.launchd_plist` helper) that writes a supervised LaunchAgent plist for `gate.py` or `python -m harness.server`.

**Extra gating beyond #909/#910/#911's generators**, because this one is materially higher-impact:
- Requires **both** `--confirm` and a non-empty `--reason` before writing anything — the script prints the full risk explanation to stderr and refuses otherwise. This mirrors `utils/personality.py`'s reason-required soul-mutation gate and `agentic/fsconnect/cli.py`'s `--confirm`-required destructive-write gate, applied to a comparably consequential action.
- `KeepAlive: {"SuccessfulExit": false}`, not a bare `KeepAlive: true` — restarts **only on crash/non-zero exit**, never after a clean operator-initiated stop. I verified this is correct (not just hoped) by reading both `gate.py`'s and `harness/server.py`'s `main()` before writing this: both delegate directly to `uvicorn.run()`, which installs its own SIGTERM handler and returns normally (exit 0) on `launchctl stop`/`bootout`. So a deliberate stop is never mistaken for a crash and silently relaunched.
- `ThrottleInterval` defaults to 30s (launchd's own default is 10s) — `gate.py`'s startup (embedding model + ChromaDB) is meaningfully slower than a lightweight poller, so a genuine crash loop stays throttled rather than hammering CPU/disk with overlapping startup attempts.
- Never calls `launchctl load`/`bootstrap` itself — prints the exact command, same as every prior generator in this series. Loading a persistent, auto-restarting listener is always left as a separate, explicit operator action.
- A **documented, not hidden** known limitation: `gate.py` has its own pre-flight port-in-use check and exits cleanly (0) if something else already owns the port — so it correctly does NOT get retried by `KeepAlive`. `harness/server.py` has no equivalent check, so a port conflict there raises inside `uvicorn.run()` and DOES crash-loop (throttled) until the port frees. This is called out explicitly in the script's own docstring and in this PR, not silently left as a surprise.

Optional `--api-key-service` reuses the same Keychain wrapper #911 introduced (duplicated onto this branch — cut from `main`, not from #911's branch, for the same independent-mergeability reasons #910's PR body explains) to inject `CYCLAW_API_KEY` without it ever being a literal plist value.

**Invariant / Governance Impact:**
- No invariant touched. `macos/generate_service_plist.py` reads `config.yaml`'s `api.port` directly via PyYAML and never imports `gate.py`, `graph.py`, or `mcp_hybrid_server.py` — confirmed by design (there was never a reason to) and by `invariant-guard`'s I6 AST check (35/35 pass, before and after).
- This does NOT touch `gate.py`, `graph.py`, `harness/server.py`, or any of their startup code. It only reads `main()` (to verify shutdown semantics) — zero lines of application code changed.

---

## Types of changes
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bugfix / Breaking change / Documentation Update / Invariant refinement

**Scope note:** Out-of-band `macos/` tooling. Zero changes to `gate.py`, `graph.py`, or `harness/`.

---

## Benefits / why

- Closes the last of the four documented gaps: "No LaunchAgent for gate.py :8787 or harness :8790 — lid close / reboot and the servers stay dead."
- The `--confirm`/`--reason` requirement means this can never be triggered accidentally or by a script that doesn't understand what it's asking for — matches this codebase's own established graduated-risk-gating idiom rather than inventing a new one.

---

## Risks to monitor (read in full before merging)

- **Not tested on real macOS** — same documented limitation as #909/#910/#911 (this is a Linux sandbox). The generated plist's structure is verified via `plistlib` round-trip and 13 dedicated tests; the actual `launchctl bootstrap`/crash-restart/`SuccessfulExit` behavior is **unverified against real launchd**. An operator with real Apple Silicon hardware should load one instance, kill -9 it, and confirm it restarts — then `launchctl stop` it and confirm it does NOT restart — before relying on this.
- **This is the one integration in the series where I'd genuinely value a second pair of eyes on the design**, not just the code, before it merges: is crash-only `KeepAlive` (vs. no `KeepAlive` at all, i.e. `RunAtLoad`-only) the right default? Is 30s the right `ThrottleInterval` floor? Is requiring `--reason` the right amount of friction, or too much/too little for how this is actually likely to be used?
- `harness/server.py`'s missing port-pre-check (noted above) means a misconfigured or double-loaded harness agent will crash-loop indefinitely at the throttle interval rather than exiting cleanly. Bounded (never faster than once per `--throttle-sec`), logged (stderr → the configured log file), but not silent-safe the way `gate.py`'s path is.
- Once loaded, this is the first CyClaw-generated artifact in the whole series that makes a *network listener* persistent across reboots — every prior integration (sync, fsconnect-trash, telegram) was a bounded, one-shot or poll-style job. That's a qualitatively different risk shape (a standing attack surface, not a periodic task) even though the loopback-only bind and existing auth model are unchanged by this PR.

---

## Checklist

- [x] Preserves all 6 invariants + I6 (invariant-guard 35/35; zero lines of `gate.py`/`graph.py`/`harness/server.py` changed)
- [x] Full sandbox validation: `GROK_API_KEY=dummy pytest tests/ -q --tb=short` green, zero failures
- [x] No new external dependencies (stdlib + PyYAML, already a project dependency)
- [x] Commit message follows the title prefix convention
- [x] `ruff check --select E,F,I,B,C4,UP,S .` clean
- [x] `python3 .claude/skills/doc-sync/doc_sync.py` — 0 drift items
- [x] `python3 .claude/skills/dep-guard/check_deps.py` — D10 still passes after adding `--cov=utils.launchd_plist`
- [x] Verified `gate.py`'s and `harness/server.py`'s actual shutdown-signal behavior by reading `main()` before choosing `KeepAlive` semantics, not by assumption

---

## Further comments

**ELI5:** Right now, CyClaw's server only runs while you have a terminal window open running it — close the window (or restart your Mac) and it's gone until you manually start it again. This adds an *optional* way to make it start automatically and restart itself if it crashes — like a service. Because that's a bigger deal than the other three integrations in this series (sync, trash cleanup, Telegram), this one makes you explicitly type out why you want it and confirm you understand what you're turning on, before it'll even write the config file — and even then, it still won't actually turn anything on until you run one more command yourself.

**Given the risk framing above, I'd suggest treating this PR differently from #909/#910/#911** — worth a closer read of the design (not just the diff) before merging, and ideally validating on real hardware first.

Verified in this Linux sandbox with the same Python 3.12.3 venv used for #908-#911.

---
_Generated by [Claude Code](https://claude.ai/code/session_01DfhYsUwMyF3dkeHjGBKa12)_

### CODE REVIEW

**Comment by cgfixit** [comment]
@codex review for security rule violations according to my project rules/invariants and security best practices overall - do a brief code scan with no internet access to limit token usage

**Comment by cgfixit** [comment]
# Consolidated macOS Manual Testing Checklist

**Scope:** Both PR #911 (Keychain wrapper + Telegram LaunchAgent) and PR #912 (gate.py/harness LaunchAgent) contain components requiring real-hardware macOS verification. This unified checklist consolidates all unverified-on-real-hardware items, enabling single-PR testing before merging both.

**Prerequisites:**
- macOS 13+ (Sonoma preferred for hardening scope)
- Python 3.12+
- `security` command available (builtin on macOS)
- Fresh CyClaw checkout on the test branches (PR #911 `claude/macos-keychain-telegram-launchagent`, PR #912 `claude/macos-gate-harness-launchagent`)
- Installed dependencies: `pip install -r requirements.txt -c constraints.txt --ignore-installed PyYAML`
- No existing launchd agents from prior CyClaw installations (or unload them first)

---

## Phase 1: Keychain Wrapper Security (PR #911)

### 1a. Bare `-w` TTY Prompt Behavior
**What to test:** `cyclaw-keychain-set.sh` uses bare `-w` flag to force `security` to prompt interactively, never passing the secret through argv.

```bash
bash macos/cyclaw-keychain-set.sh com.cgfixit.test-token-123
# Expected: Interactive prompt, no secret in ps/procfs output
```

### 1b. TTY Requirement (Non-Interactive Failure)
```bash
echo "test-secret" | bash macos/cyclaw-keychain-set.sh com.cgfixit.test-pipe
# Expected exit code: 1, stderr mentions TTY requirement
```

### 1c & 1d. Trust Pinning & ACL Attribution
- Verify items stored with `-T /usr/bin/security` (use `security dump-keychain` or Keychain Access.app)
- Verify ACL attributes grant to `/usr/bin/security`, NOT `/bin/bash`
- Test via launchd wrapper: `cyclaw-keychain-env.sh` should fetch without prompts (ACL already set)

---

## Phase 2: Telegram Plist Generation (PR #911)

### 2a. Poll Plist: Keychain Wrapping for Bot Token + API Key
```bash
python macos/generate_service_plist.py --service telegram-poll --reason "test" --confirm
# Verify: chained Keychain wrappers in ProgramArguments, no secrets in plist
defaults read ~/Library/LaunchAgents/com.cgfixit.cyclaw.telegram-poll.plist ProgramArguments
```

### 2b. Health Plist: Single Keychain Wrapper
```bash
python macos/generate_service_plist.py --service telegram-health --reason "test" --confirm
# Verify: single Keychain wrapper for bot token only
```

### 2c. Custom Env Var Names
- Edit config.yaml to use custom `bot_token_env` and `query.api_key_env` names
- Verify plists use custom names in wrapper chain
- Restore config.yaml

---

## Phase 3: Gate LaunchAgent (PR #912)

### 3a. Plist Structure
```bash
python macos/generate_service_plist.py --service gate --reason "test" --confirm
# Verify: Label, RunAtLoad=true, KeepAlive={SuccessfulExit:false}, ThrottleInterval=30
```

### 3b. Port Conflict Behavior (Clean Exit)
- Start gate manually on configured port
- Load LaunchAgent; verify it does NOT crash-loop (gate's pre-check exits cleanly)
- Stop manual gate, verify LaunchAgent can load

### 3c. Clean Stop (SIGTERM)
- Load gate, verify running
- `launchctl stop com.cgfixit.cyclaw.gate`, verify process exits
- Verify KeepAlive does NOT restart it (clean stop, not crash)

### 3d. Crash-Loop Throttle
- Inject intentional crash in gate.py
- Load agent, monitor restart timing
- Verify restarts spaced ≥30s apart (ThrottleInterval honored)
- Cleanup: remove crash, bootout agent

### 3e. Config File (No --config Flag)
- Verify plist has no `--config` argument
- Verify gate reads from repo's config.yaml

---

## Phase 4: Harness LaunchAgent (PR #912)

### 4a. Plist Structure
```bash
python macos/generate_service_plist.py --service harness --reason "test" --confirm
# Verify: Label, RunAtLoad=true, EnvironmentVariables with CYCLAW_HOME/CYCLAW_REPO
```

### 4b. Clean Stop
- Load harness, verify running
- `launchctl stop com.cgfixit.cyclaw.harness`
- Verify stopped and NOT restarted

### 4c. Port Conflict Crash-Loop
- Start listener on port 8790
- Load harness, verify crash-loop (harness has no pre-check)
- Verify throttled by ThrottleInterval
- Cleanup

---

## Phase 5: Installation/Uninstallation

### 5a. `macos/install-cyclaw.sh` Integration
- Check: script calls `generate_service_plist.py` with `--confirm` and `--reason`
- Check: only for gate and harness (telegram agents optional)

### 5b. Uninstall Cleanup
- Run `macos/uninstall-cyclaw.sh` (if exists)
- Verify LaunchAgent plists removed

---

## Phase 6: Edge Cases

### 6a. Confirm + Reason Gating
```bash
python macos/generate_service_plist.py --service gate --reason "test"      # Missing --confirm → exit 1
python macos/generate_service_plist.py --service gate --confirm            # Missing --reason → exit 1
python macos/generate_service_plist.py --service gate --confirm --reason "   "  # Blank → exit 1
```

### 6b. Non-Darwin Rejection
- Verify script refuses to run on non-macOS systems (exit code 3)

### 6c. Config File Validation
- Move config.yaml, run script → exit code 3
- Restore config.yaml

---

## Verification Checklist

- [ ] Phase 1: Keychain wrapper security (all 4 items)
- [ ] Phase 2: Telegram plist generation (all 3 items)
- [ ] Phase 3: Gate LaunchAgent (all 5 items)
- [ ] Phase 4: Harness LaunchAgent (all 3 items)
- [ ] Phase 5: Installation/uninstallation workflows (all 2 items)
- [ ] Phase 6: Edge cases and hardening (all 3 items)

---

## Known Limitations (Document, Don't Fix)

1. **Keychain ACL prompt:** The "Allow / Always Allow" dialog attribution to `/usr/bin/security` vs. `/bin/bash` has not been independently verified against the actual prompt shown on real hardware. Follows Apple's published Security framework model.

2. **KeepAlive vs. Clean Stop:** Assumes uvicorn/gate.py/harness exit cleanly on SIGTERM (verified by code inspection).

3. **Sandboxing / TCC:** CyClaw uses no TCC-gated entitlements. Future cron-based launchd may require TCC grants — out of scope.

4. **Signed / Notarized Distribution:** Not in scope; suitable for personal/operator-driven installation only.

---

## Summary

This checklist verifies all unverified-on-real-hardware components across both PRs in a single test run. PR #911 focuses on Keychain and Telegram integration; PR #912 focuses on gate.py and harness process supervision. Both rely on macOS-specific APIs and launchd semantics that cannot be fully tested in CI. Completion of this checklist clears both PRs for merge.

**Next step:** PR #911 can be merged once CI passes. Run this checklist against both branches before merging #912 (or merge #911 first, then test both features together from the consolidated #912 testing).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: macos/cyclaw-keychain-set.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+read -r -s SECRET
+printf '\n'
+
+if [ -z "$SECRET" ]; then
+  echo "cyclaw-keychain-set: empty secret refused" >&2
+  exit 1
+fi
+
+ACCOUNT="$(id -un)"
+security add-generic-password -a "$ACCOUNT" -s "$SERVICE" -w "$SECRET" -U
```
[the remark this answers is not in the document]
Comment (cgfixit) [reply macos/cyclaw-keychain-set.sh]: Fixed. `cyclaw-keychain-set.sh` now calls `security add-generic-password ... -w` with a **bare** `-w` (no attached value) instead of `-w "$SECRET"`. `security`'s `-w` takes an optional attached value (getopt `w::` style) — bare `-w` makes `security` itself prompt for the password via a secure TTY read (`readpassphrase(3)`, no echo), so the secret is never a shell variable in this script and never an argv token of the `security` child process. A TTY check (`[ ! -t 0 ]`) fails closed when run non-interactively. Verified end-to-end with a real `pty.openpty()`-based test (`test_set_script_stores_via_pty_and_never_echoes_secret_to_argv`) that confirms the secret never appears in the logged argv.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: macos/generate_service_plist.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+
+def _maybe_print_harness_api_key_note(args: argparse.Namespace) -> None:
+    if args.service != "harness" or not args.api_key_service:
+        return
+    print()
+    print("  NOTE: harness/server.py does not currently read CYCLAW_API_KEY --")
+    print("  --api-key-service has no effect for --service harness today. The")
+    print("  Keychain wrapper still runs and exports the variable, but nothing")
+    print("  in the harness process consumes it.")
```
[the remark this answers is not in the document]
Comment (cgfixit) [reply macos/generate_service_plist.py]: Fixed. The false "no effect" warning for `--service harness --api-key-service ...` has been removed — you're right that `harness/server.py` guards its routes with `require_api_key`, which reads `CYCLAW_API_KEY` via `utils/auth.py`, so the injected key is exactly what makes the supervised harness usable. Regression test `test_harness_plist_with_api_key_service_wraps_keychain` (renamed from the old test asserting the incorrect warning) now asserts the Keychain-wrapper chain is present with no misleading note, and `test_gate_plist_with_api_key_service_has_no_no_effect_note` locks the "no effect" string's absence.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: macos/generate_service_plist.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    if args.throttle_sec <= 0:
+        print("error: --throttle-sec must be > 0", file=sys.stderr)
+        return 2
+
+    label = _LABELS[args.service]
+    env: dict[str, str] = {}
+
+    if args.service == "gate":
+        port = _read_gate_port(Path(args.config).resolve())
+        inner_argv = [launchd_plist.python_executable(), str(_REPO_ROOT / "gate.py")]
```
[the remark this answers is not in the document]
Comment (cgfixit) [reply macos/generate_service_plist.py:169]: Fixed. `_maybe_print_gate_config_mismatch()` now prints a loud, explicit warning whenever `--service gate --config <path>` is passed a path that differs from the repo-root `config.yaml` gate.py actually loads at runtime — spelling out that the reported port is not necessarily what the supervised process will bind to, and that `config.yaml` must be edited directly to change it. That function already existed on this branch but had no regression test locking the behavior; just pushed `f16596a` adding `test_gate_plist_custom_config_warns_of_port_mismatch` (asserts the warning fires and names the mismatched path) and `test_gate_plist_default_config_has_no_mismatch_warning` (asserts silence on the no-mismatch path). 17/17 green, `ruff` clean, `invariant-guard` 35/35.

---
_Generated by [Claude Code](https://claude.ai/code)_
